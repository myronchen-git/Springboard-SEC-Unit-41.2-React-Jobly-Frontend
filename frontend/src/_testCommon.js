const companies = Object.freeze([
  Object.freeze({
    handle: 'comp1',
    name: 'Company 1',
    description: 'Company 1 description.',
    numEmployees: 12,
    logoUrl: 'logo1.png',
  }),
  Object.freeze({
    handle: 'comp2',
    name: 'Company 2',
    description: 'Company 2 description.',
    numEmployees: 123,
    logoUrl: 'logo2.png',
  }),
  Object.freeze({
    handle: 'comp3',
    name: 'Company 3',
    description: 'Company 3 description.',
    numEmployees: 3,
  }),
  Object.freeze({
    handle: 'comp4',
    name: 'Company 4',
    description: 'Company 4 description.',
    logoUrl: 'logo4.png',
  }),
]);

const jobs = Object.freeze([
  Object.freeze({
    id: '1',
    title: 'Job 1',
    salary: 1,
    equity: 0.1,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
  Object.freeze({
    id: '2',
    title: 'Job 2',
    salary: 2,
    equity: 0.2,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
]);

const comp1Jobs = [];
[jobs[0], jobs[1]].forEach((job) => {
  const condensedJob = { ...job };
  delete condensedJob.companyHandle;
  delete condensedJob.companyName;
  comp1Jobs.push(Object.freeze(condensedJob));
});

const companyDetails = Object.freeze({
  handle: 'comp1',
  name: 'Company 1',
  description: 'Company 1 description.',
  numEmployees: 12,
  logoUrl: 'logo1.png',
  jobs: Object.freeze(comp1Jobs),
});

// ==================================================

export { companies, companyDetails, jobs };
