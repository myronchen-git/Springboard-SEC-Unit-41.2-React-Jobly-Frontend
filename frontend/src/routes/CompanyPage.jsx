import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

import JoblyApi from '../../../api';

// ==================================================

/**
 * Displays detailed info about a company and a list of its job openings.
 *
 * @param {Object} props - React component properties.
 * @param {Array} props.applications - A list of job IDs that were applied to.
 */
function CompanyPage({ applications }) {
  const { handle } = useParams();
  const [company, setCompany] = useState({});
  const [jobs, setJobs] = useState([]);

  useEffect(() => {
    JoblyApi.getCompany(handle).then((companyData) => {
      setJobs([...companyData.jobs]);
      delete companyData.jobs;
      setCompany(companyData);
    });
  }, [handle]);

  return (
    <main className="CompanyPage">
      <section className="CompanyPage__info">
        <header>
          <img src={company?.logoUrl} alt={`${company.name} Logo`} />
          <h1>{company.name}</h1>
        </header>
        {company.numEmployees && <p>{company.numEmployees} employees</p>}
        <p>{company.description}</p>
      </section>
      <ul className="CompanyPage__jobs">
        {jobs.map((job) => (
          <li key={job.id}>
            <p>{job.title}</p>
          </li>
        ))}
      </ul>
    </main>
  );
}

// ==================================================

CompanyPage.propTypes = {
  applications: PropTypes.array.isRequired,
};

export default CompanyPage;
