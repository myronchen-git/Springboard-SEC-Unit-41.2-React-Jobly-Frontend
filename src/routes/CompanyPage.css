.CompanyPage__info header {
  margin: 1em auto;
}

.CompanyPage__logo-container {
  margin: 1em;
}

.CompanyPage__logo {
  height: 100px;
  width: 100px;
}
