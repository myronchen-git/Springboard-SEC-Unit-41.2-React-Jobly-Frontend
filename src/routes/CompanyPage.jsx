import { useContext, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { Alert, List } from 'reactstrap';

import JoblyApi from '../api.js';
import JobCard from '../components/JobCard.jsx';
import { UserContext } from '../contexts.jsx';

import './CompanyPage.css';

// ==================================================

/**
 * Displays detailed info about a company and a list of its job openings.
 */
function CompanyPage() {
  const { handle } = useParams();
  const [company, setCompany] = useState({});
  const [jobs, setJobs] = useState([]);
  const { applications } = useContext(UserContext);
  const [errorMessage, setErrorMessage] = useState();

  useEffect(() => {
    JoblyApi.getCompany(handle)
      .then((companyData) => {
        setJobs([...companyData.jobs]);
        delete companyData.jobs;
        setCompany(companyData);
      })
      .catch((err) => setErrorMessage(err[0]));
  }, [handle]);

  return (
    <main className="CompanyPage">
      {errorMessage ? (
        <Alert color="danger">{errorMessage}</Alert>
      ) : (
        <>
          <section className="CompanyPage__info">
            <header>
              <div className="CompanyPage__logo-container">
                <img
                  className="CompanyPage__logo"
                  src={company?.logoUrl}
                  alt={`${company.name} Logo`}
                />
              </div>
              <h1>{company.name}</h1>
            </header>
            {company.numEmployees && <p>{company.numEmployees} employees</p>}
            <p>{company.description}</p>
          </section>
          <List className="CompanyPage__jobs" type="unstyled">
            {jobs.map((job) => (
              <li key={job.id}>
                <JobCard job={job} isApplied={applications.includes(job.id)} />
              </li>
            ))}
          </List>
        </>
      )}
    </main>
  );
}

// ==================================================

export default CompanyPage;
