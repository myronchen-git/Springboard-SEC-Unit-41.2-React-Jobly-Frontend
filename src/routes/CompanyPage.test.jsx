import { act, fireEvent, render } from '@testing-library/react';
import { useParams } from 'react-router-dom';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { companyDetails } from '../_testCommon.js';
import JoblyApi from '../api.js';
import { UserContext } from '../contexts.jsx';
import CompanyPage from './CompanyPage.jsx';

// ==================================================

vi.mock('react-router-dom');

vi.mock(import('../api.js'), () => {
  const MockJoblyApi = vi.fn();
  MockJoblyApi.getCompany = vi.fn();

  return {
    default: MockJoblyApi,
  };
});

// ==================================================

describe('CompanyPage', () => {
  const applications = Object.freeze(['2', '3']);
  const mockApplyToJob = vi.fn();

  beforeEach(() => {
    mockApplyToJob.mockReset();

    useParams.mockReturnValue({ handle: companyDetails.handle });
    JoblyApi.getCompany.mockResolvedValue(structuredClone(companyDetails));
  });

  it('renders.', async () => {
    await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <CompanyPage />
        </UserContext.Provider>
      )
    );
  });

  it('matches snapshot.', async () => {
    const { asFragment } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <CompanyPage />
        </UserContext.Provider>
      )
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays data about company.', async () => {
    // Act
    const { getByAltText, getByText } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <CompanyPage />
        </UserContext.Provider>
      )
    );

    // Assert
    expect(getByAltText(`${companyDetails.name} Logo`)).toBeVisible();
    expect(getByText(companyDetails.name)).toBeVisible();
    expect(getByText(companyDetails.description)).toBeVisible();
    expect(getByText(companyDetails.numEmployees + ' employees')).toBeVisible();

    companyDetails.jobs.forEach((job) => {
      expect(getByText(job.title)).toBeVisible();
    });
  });

  it('displays the correct apply button status.', async () => {
    // Act
    const { getAllByText } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <CompanyPage />
        </UserContext.Provider>
      )
    );

    // Assert
    // Fragile!
    const applybtns = getAllByText('Appl', { exact: false });

    expect(applybtns[0]).toHaveTextContent('Apply');
    expect(applybtns[1]).toHaveTextContent('Applied');
  });

  it('allows applying to a job.', async () => {
    // Arrange
    const { queryAllByText } = await act(async () =>
      render(
        <UserContext.Provider
          value={{ applications, applyToJob: mockApplyToJob }}
        >
          <CompanyPage />
        </UserContext.Provider>
      )
    );

    const applyBtns = queryAllByText('Apply');

    // Act
    await act(async () => fireEvent.click(applyBtns[0]));

    // Assert
    expect(mockApplyToJob).toHaveBeenCalledOnce();
    expect(mockApplyToJob).toHaveBeenCalledWith(companyDetails.jobs[0].id);
  });
});
