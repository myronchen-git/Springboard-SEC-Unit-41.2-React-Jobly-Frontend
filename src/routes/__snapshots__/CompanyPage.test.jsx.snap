// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompanyPage > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="CompanyPage"
  >
    <section
      class="CompanyPage__info"
    >
      <header>
        <div
          class="CompanyPage__logo-container"
        >
          <img
            alt="Company 1 Logo"
            class="CompanyPage__logo"
            src="logo1.png"
          />
        </div>
        <h1>
          Company 1
        </h1>
      </header>
      <p>
        12 employees
      </p>
      <p>
        Company 1 description.
      </p>
    </section>
    <ul
      class="CompanyPage__jobs list-unstyled"
    >
      <li>
        <div
          class="JobCard card"
        >
          <div
            class="JobCard__info card-body"
          >
            <h4
              class="card-title"
            >
              Job 1
            </h4>
            <p
              class="card-text"
            >
              Salary: $1
            </p>
            <p
              class="card-text"
            >
              Equity: 0.1
            </p>
          </div>
          bound HTMLFormElement {
            "0": <button
              class="btn btn-secondary"
              type="button"
            >
              Apply
            </button>,
          }
        </div>
      </li>
      <li>
        <div
          class="JobCard card"
        >
          <div
            class="JobCard__info card-body"
          >
            <h4
              class="card-title"
            >
              Job 2
            </h4>
            <p
              class="card-text"
            >
              Salary: $2
            </p>
            <p
              class="card-text"
            >
              Equity: 0.2
            </p>
          </div>
          bound HTMLFormElement {
            "0": <button
              class="btn btn-secondary disabled"
              disabled=""
            >
              Applied
            </button>,
          }
        </div>
      </li>
    </ul>
  </main>
</DocumentFragment>
`;
